Change package testing to use RuntimeTargets rather than RID-specific restore

ericstj opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/2017

@Anipik and I were playing around with this today and realized we were very close to enabling package testing using just single restore.  I went ahead and finished our prototype.  This should greatly reduce the restore cost and total duration of package testing.

@ViktorHofer this should unblock #53439

ericstj committed:
Change package testing to use RuntimeTargets rather than RID-specific restore

ericstj commented on src/libraries/pkg/test/packageTest.targets:
```diff
@@ -71,11 +70,36 @@
+      <RuntimeLibToTestDependency Include="@(ReferencePath)" Condition="'%(ReferencePath.NuGetPackageId)' != '$(TestPackageId)'" />
+
+      <!-- Some dependent packages may be excluded from compile, consider these as candiates as well.  
+           We assume they must expose a RID agnostic asset, thus we only check RuntimeCopyLocalItems. -->
```
To be more correct / replicate the actual runtime behavior here we would need to replicate the host selection logic to test with the right runtime assets for dependencies.  It's not so interesting to do this type of testing since the SDK doesn't make it possible for a library to reference a runtime specific asset of another library or project, and we don't typically do this with our custom config system.  It's bad practice for library to depend on implementation-only publics of another library; if one library needed it why wouldn't a customer library need it?

So the asset analysis ends up looking like this:
1. Package under test: Foreach runtime asset group
2. Package dependencies w/ exclude=comple: RID-less runtime asset group
3. Package dependencies w/o exclude=compile: compile asset group
4. Framework: ref-pack / reference assemblies only.

Note that 2-4 are changes from what we previously did, but these are "safe" changes, since if these validations succeed then API compat will ensure that transitively things should succeed.

ericstj committed:
Merge remote-tracking branch 'upstream/main' into packageTestRuntimeTargets

ericstj committed:
Upload binlogs for package testing

ericstj commented:
I broke this in a hilarious way.  The structure of the package testing projects used to be
```
{packageId}\{TFM}\project.csproj
```
I removed the TFM since I can make the projects cross-target now that they don't need RIDs
```
{packageId}\project.csproj
```
When hexlix runs these, it's removing the packageId folder and placing the project in the root.

In the same directory as the `packages` folder.  So it looks like this:
```
{workItemDir}\project.csproj
{workItemDir}\packages\...
```

Then what happens is all the package content gets globbed as `None`, and passed in as `CandidateAssemblyFiles` to RAR.  CandidateAssemblyFiles happen to be considered first in RAR's precedence when resolving assemblies.  So we were getting seemingly random assemblies resolved, based on whatever appeared first in the package cache (often not even appropriate for the framework).

Guess I need to restructure the helix test directories for these 😆.

ericstj commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -80,6 +80,7 @@
 
   <PropertyGroup Condition="'$(TestPackages)' == 'true'">
     <HelixCommand>dotnet msbuild %HELIX_CORRELATION_PAYLOAD%\test.msbuild</HelixCommand>
+    <HelixCommand>$(HelixCommand) /bl:%HELIX_WORKITEM_UPLOAD_ROOT%\msbuild.binlog</HelixCommand>
```
I'm curious what folks thoughts are on this.  Should I keep it in? Seems similar to dumps.  Ideally we only upload if there is a failure though.

ViktorHofer replied:
Yes please keep it. We always generate binlogs in CI as remoting onto a machine is often not possible when diagnosing failures in CI.

joperezr replied:
I agree keeping them is good, that said I probably would rename it to something like packageTesting.binlog perhaps?

MattGal replied:
Looking around at things like Xharness and Blazor, lots of work items are invariably uploading binlogs of 40-700 K every day so I don't think this really needs to even consider being conditional unless this is going to be running on on-prem machines where storage pricing changes the equation a bit.    +1 to Joe's comment about a more helpful name.

ericstj replied:
Yep, I changed the name to match the work-item friendly name + package.test.binlog.  That way it behaves better when downloaded. (I think I'm currently at "msbuild (80).binlog" in my downloads folder). 😆

ViktorHofer commented:
As an alternative, you could also disable the globbing.

ViktorHofer commented:
As you are removing some of the frameworkSettings files, why not remove the portable and netcore50 as well?

Anipik commented on src/libraries/pkg/test/testPackages.proj:
```diff
@@ -140,15 +123,14 @@
       <_packageVersion>%(SupportedPackage.Version)</_packageVersion>
-      <_projectTFM>%(SupportedPackage.TargetFrameworkShort)</_projectTFM>
-      <_projectRIDs>%(SupportedPackage.RuntimeIdentifiers)</_projectRIDs>
+      <_projectTFMs>@(SupportedPackage->'%(TargetFrameworkShort)')</_projectTFMs>
```
Why are we using tfms here instead of tfm ?

ericstj replied:
By eliminating RIDs I'm able to use a single project for testing.   RIDs required separate projects per TFM since they had to set different RIDs per TFM.  This improves performance of the tests and simplifies the process.

Anipik replied:
would it make sense to just make arcade task return tfms object here ?

ericstj commented:
> As you are removing some of the frameworkSettings files, why not remove the portable and netcore50 as well?

I'm removing the ones that were no longer needed due to the change I made, since they were related to using the RuntimeIdentifier, or testing against runtime assemblies.  I'll consider cleaning others if they're clearly dead code, but I don't want to feature creep this PR.

> As an alternative, you could also disable the globbing.

I don't think that's necessary.  Better to keep projects simple rather than mess with them to accommodate a wacky thing that customers should never do (putting packages restore folder under project).

Anipik commented on src/libraries/pkg/test/packageTest.targets:
```diff
@@ -105,55 +132,27 @@
-    </_testDependsOn>
+    </TestDependsOn>
+    <TestDependsOn Condition="'$(TargetFramework)' == ''">
+      _ComputeTargetFrameworkItems
```
nice we are using the cross targeting to iterate over the tfms.

ericstj replied:
I debated just setting `<InnerTargets>Test<InnerTargts>` and this would make build just do the testing.  That would avoid use of private target and item.  Works well for these test-specific projects.  Thoughts on that?

Anipik replied:
yes, i think that would be best. IIRC our build in the current form is already hooked up to Test target.

Anipik commented on src/libraries/pkg/test/packageTest.targets:
```diff
@@ -105,55 +132,27 @@
       VerifyRuntimeTypes;
-    </_testDependsOn>
+    </TestDependsOn>
+    <TestDependsOn Condition="'$(TargetFramework)' == ''">
```
should we use ```Condition="'$(IsCrossTargetingBuild)' == 'true'"```  instead ?

ericstj replied:
Will do this based on what we decide for https://github.com/dotnet/runtime/pull/53575#discussion_r644113768

joperezr commented on src/libraries/pkg/test/frameworkSettings/netcoreapp1.0/settings.targets:
```diff
@@ -1,32 +0,0 @@
-<Project>
-  <ItemGroup Condition="'$(RuntimeIdentifier)' != ''">
```
I'm not sure I follow Why these can be removed.

ericstj replied:
Because we no longer test with `RuntimeIdentifier` at all.

ericstj commented:
>> As an alternative, you could also disable the globbing.

> I don't think that's necessary. Better to keep projects simple rather than mess with them to accommodate a wacky thing that customers should never do (putting packages restore folder under project).

So it turns out we can even delete the custom packages folder as Helix already does this for us:
https://github.com/dotnet/arcade/blob/1b17d7c9f0c2100bac5766b154c99fb31ab20eb4/src/Microsoft.DotNet.Helix/Sdk/tools/dotnet-cli/DotNetCli.targets#L21-L22

I'm looking into keeping the binlog, but only uploading it when there is a failure.

ericstj committed:
Address feedback

Simplify outer build of package tests to just use InnerTargets extension point.

Use Helix's work-item isolated packages folder

Give a better name to binlogs

ericstj committed:
Cleanup old suppressions

ericstj committed:
Use ReferenceCopyLocalPaths for runtime testing

This item will be conflict-resolved whereas RuntimeCopyLocalItems was not.

Also fix the case where a package intentionally provides no assets (yet installs)

ericstj committed:
Don't include runtime dependencies when reference was excluded by conflict resolution

ericstj commented:
PKCS test is failing due to https://github.com/dotnet/sdk/issues/18129.  I'm planning to change how I did the workaround for this.  Previously I was relying on conflict resolution dropping a compile asset to indicate what package references should not be considered when evaluating closure of runtimes, but this doesn't work if the packages that should have been dropped were Exclude=compile.  I'm either going to feed all the ref-pack assets into conflict resolution for consideration at runtime as well, or I'll leverage the `PackageOverrides` mechanism if I can get it to work.

ericstj committed:
Fix import conventions in outer build

ericstj committed:
Workaround conflict resolution on .NETStandard runtime

.NETStandard doesn't conflict resolve runtime assets.
Workaround by feeding .NETStandard references as runtime for purposes of conflict resolution.

safern reviewed (commented):
LGTM, thanks!
